src/modules/wallet/utils.ts

```diff
@@ -83,57 +84,85 @@ export async function getTargetNetworkProvider(chainId: ChainId) {
+  ERROR = 'error',
+  SUCCESS = 'success'
+}
+
+export type TransactionEventData<T extends TransactionEventType> = {
+  type: T
+} & (T extends TransactionEventType.ERROR
+  ? { error: Error }
+  : T extends TransactionEventType.SUCCESS
+  ? { txHash: string }
+  : {})
+
+export const transactionEvents = new EventEmitter()
+
 export async function sendTransaction(
   contract: ContractData,
   getPopulatedTransaction: (
     populateTransaction: Contract['populateTransaction']
   ) => Promise<PopulatedTransaction>
 ) {
-  // get connected provider
-  const connectedProvider = await getConnectedProvider()
-  if (!connectedProvider) {
-    throw new Error('Provider not connected')
-  }
+  try {
+    // get connected provider
+    const connectedProvider = await getConnectedProvider()
+    if (!connectedProvider) {
+      throw new Error('Provider not connected')
+    }
 
-  // get current chain id
-  const chainIdHex = await connectedProvider.request({
-    method: 'eth_chainId',
-    params: []
-  })
-  const chainId = parseInt(chainIdHex as string, 16)
-
-  // get a provider for the target network
-  const targetNetworkProvider = await getTargetNetworkProvider(contract.chainId)
-
-  // intantiate the contract
-  const contractInstance = new Contract(
-    contract.address,
-    contract.abi,
-    targetNetworkProvider
-  )
-
-  // populate the transaction data
-  const unsignedTx = await getPopulatedTransaction(
-    contractInstance.populateTransaction
-  )
-
-  // if the connected provider is in the target network, use it to sign and send the tx
-  if (chainId === contract.chainId) {
-    const signer = targetNetworkProvider.getSigner()
-    const tx = await signer.sendTransaction(unsignedTx)
-    return tx.hash
-  } else {
-    // otherwise, send it as a meta tx
-    const hash = await sendMetaTransaction(
-      connectedProvider,
-      targetNetworkProvider,
-      unsignedTx.data!,
-      contract,
-      {
-        serverURL: getTransactionsApiUrl()
-      }
+    // get current chain id
+    const chainIdHex = await connectedProvider.request({
+      method: 'eth_chainId',
+      params: []
+    })
+    const chainId = parseInt(chainIdHex as string, 16)
+
+    // get a provider for the target network
+    const targetNetworkProvider = await getTargetNetworkProvider(
+      contract.chainId
     )
-    return hash
+
+    // intantiate the contract
+    const contractInstance = new Contract(
+      contract.address,
+      contract.abi,
+      targetNetworkProvider
+    )
+
+    // populate the transaction data
+    const unsignedTx = await getPopulatedTransaction(
+      contractInstance.populateTransaction
+    )
+
+    // if the connected provider is in the target network, use it to sign and send the tx
+    if (chainId === contract.chainId) {
+      const signer = targetNetworkProvider.getSigner()
+      const tx = await signer.sendTransaction(unsignedTx)
+      transactionEvents.emit(TransactionEventType.SUCCESS, { txHash: tx.hash })
```

Comment:
I'm not sure if we should make sendTransaction a sagas handler now that we need to impact the state to show the toasts.